File: C++/1069-the_black_hole_of_numbers.cc
#include <iostream>
#include <string>
#include <algorithm>


// Pad a numeric string with zeros on the left in-place,
// to fill a field of the given width.
void zfill(std::string &s, std::size_t width) {
    while (s.size() != width) {
        s = '0' + s;
    }
}


int main() {

    std::string num;
    std::cin >> num;
    zfill(num, 4);
    std::string num_r(num.crbegin(), num.crend());
    if (num == num_r) {
        std::cout << num << " - " << num << " = 0000\n";
    } else {
        while (true) {
            std::sort(num.rbegin(), num.rend());
            std::string num_r(num.crbegin(), num.crend());
            auto res = std::to_string(std::stoi(num) - std::stoi(num_r));
            zfill(res, 4);
            std::cout << num << " - " << num_r << " = " << res << '\n';
            if (res == "6174") break;
            else num = res;
        }
    }

    return 0;
}

File: Python/1069-the_black_hole_of_numbers.py


if __name__ == '__main__':
    num = input().zfill(4)
    if num == num[::-1]:
        print(f"{num} - {num} = 0000")
    else:
        while True:
            num = ''.join(sorted(num, reverse=True))
            num_r = num[::-1]
            res = str(int(num) - int(num_r)).zfill(4)
            print(f'{num} - {num_r} = {res}')
            if res == '6174':
                break
            else:
                num = res
